environs/manual/winrmprovisioner/winrmprovisioner.go

```diff
@@ -0,0 +1,581 @@
+		Secure:   false,
+	}); err != nil {
+		return errors.Annotatef(err, "cannot create a new http winrm client ")
+	}
+
+	logger.Infof("Trying http client as user %s on %s", args.Host, args.User)
+	if err = args.WClient.Ping(); err != nil {
+		logger.Debugf("WinRM unsecure listener is not enabled on %s", args.Host)
+		return errors.Annotatef(err, "cannot provision, because all winrm default connections failed")
+	}
+
+	defClient := args.WClient //save http client here in case we need to revert back
+	logger.Infof("Trying to enable https client certificate authenticaion")
+	if args.WClient, err = enableCertAuth(args); err != nil {
+		logger.Infof("Cannot enable client auth cert authentication for winrm")
+		logger.Infof("Reverting back to usecure client interaction")
+		// everything failed expept the http clinet
+		// so use this to provision the machine
+		args.WClient = defClient
+		return nil
+	}
+	logger.Infof("Client certs are installed and setup on the %s with user %s", args.Host, args.User)
+
+	if err = args.WClient.Ping(); err == nil {
+		logger.Infof("Winrm client authentication is now enabled on the host %s with user %s", args.Host, args.User)
+		return nil
+	}
+
+	logger.Infof("Winrm https connection is broken, can't retrive a response")
+	logger.Infof("Reverting back to usecure client interactions")
+	// everything failed expept the http clinet
+	// so use this to provision the machine
+	args.WClient = defClient
+
+	return nil
+
+}
+
+// enableCertAuth enables https cert auth interactions
+// with the winrm listener and returns the client
+func enableCertAuth(args *manual.ProvisionMachineArgs) (manual.WinrmClientAPI, error) {
+	var stderr bytes.Buffer
+	pass := args.WClient.Password()
+
+	scripts, err := bindInitScripts(pass, args.WKeys)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	for _, script := range scripts {
+		err = args.WClient.Run(script, args.Stdout, &stderr)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if stderr.Len() > 0 {
+			return nil, errors.Errorf(
+				"encountered error executing cert auth script: %s",
+				stderr.String(),
+			)
+		}
+	}
+
+	cfg := winrm.ClientConfig{
+		User:    args.User,
+		Host:    args.Host,
+		Key:     args.WKeys.ClientKey(),
+		Cert:    args.WKeys.ClientCert(),
+		Timeout: 25 * time.Second,
+		Secure:  true,
+	}
+
+	caCert := args.WKeys.CACert()
+	if caCert == nil {
+		logger.Infof("Skipping winrm CA validation")
+		cfg.Insecure = true
+	} else {
+		cfg.CACert = caCert
+	}
+
+	return winrm.NewClient(cfg)
+}
+
+// bindInitScripts creates a series of scripts in a standard
+// (utf-16-le, base64)format for passing to the winrm conn to be executed remotely
+// we are doing this instead of one big script because winrm supports
+// just 8192 length commands. We know we have an amount of prefixed scripts
+// that we want to bind for the init process so create an array of scripts
+func bindInitScripts(pass string, keys *winrm.X509) ([3]string, error) {
```

Comment:
Yeah it ranges through the result. But I choosed a simple array over slice for many reasons.
Why should I assume or let the compiler decide if the space is stored on **heap** or **stack**?
It's more a memory efficient way to just make a simple array on stack and use it. It's not like the array needs to get bigger or modify it any time soon.
Why should I clutter the code with ```append``` ? Or use unnecessary call to ```make``` at the start ?(And ```make``` it's not 100% it will make the decision to make it on **heap**)
Using a simple array I inform the caller that he/she will deal with a constant number of scripts.